src/main/java/uk/gov/hmcts/ccd/domain/service/DocLinksRestoreService.java

```diff
@@ -0,0 +1,134 @@
+        return findInOldCasesBasedOnEventHistory(query.getResultList());
+    }
+
+    private List<CaseDetailsEntity> findDocLinksMissedNewCases(List<String> jurisdictionList) {
+        Query query = jurisdictionList.isEmpty() ? em.createNativeQuery(NEW_CASE_QUERY, CaseDetailsEntity.class) :
+            em.createNativeQuery(NEW_CASE_QUERY_WITH_JUIDS, CaseDetailsEntity.class).setParameter("jids", jurisdictionList);
+        return findInNewCasesBasedOnEventHistory(query.getResultList());
+    }
+
+    private List<CaseDetailsEntity> findInOldCasesBasedOnEventHistory(List<CaseDetailsEntity> caseDetailsEntities) {
+        Map<Long, CaseDetailsEntity> caseMap = getCaseIdMap(caseDetailsEntities);
+
+        Query query = em.createNativeQuery(LAST_EVENT_WITH_VALID_DATA, CaseAuditEventEntity.class).setParameter("caseIds", caseMap.keySet());
+        List<CaseAuditEventEntity> lastWellKnownEvents = query.getResultList();
+
+        Map<CaseDetailsEntity, CaseAuditEventEntity> caseToEventsMap = lastWellKnownEvents.stream()
+            .collect(Collectors.toMap(event -> caseMap.get(event.getCaseDataId()), Function.identity()));
+
+        return caseToEventsMap.keySet().stream()
+            .filter(key -> hasMissingDocuments(key, caseToEventsMap.get(key)))
+            .collect(Collectors.toList());
+    }
+
+    private List<CaseDetailsEntity> findInNewCasesBasedOnEventHistory(List<CaseDetailsEntity> caseDetailsEntities) {
+        Map<Long, CaseDetailsEntity> caseMap = getCaseIdMap(caseDetailsEntities);
+
+        Query query = em.createNativeQuery(NEW_CASE_EVENTS, CaseAuditEventEntity.class).setParameter("caseIds", caseMap.keySet());
+        List<CaseAuditEventEntity> caseEvents = query.getResultList();
+
+        Map<Long, List<CaseAuditEventEntity>> eventsByCaseIdMap = caseEvents.stream().collect(groupingBy(CaseAuditEventEntity::getCaseDataId));
+
+        Map<CaseDetailsEntity, List<CaseAuditEventEntity>> caseToEventsMap = caseDetailsEntities.stream()
+            .collect(Collectors.toMap(Function.identity(), caseData -> eventsByCaseIdMap.get(caseData.getId())));
+
+        return caseToEventsMap.keySet().stream()
+            .filter(key -> hasMissingDocuments(key, caseToEventsMap.get(key)))
+            .collect(Collectors.toList());
+    }
+
+    private Map<Long, CaseDetailsEntity> getCaseIdMap(List<CaseDetailsEntity> caseDetailsEntities) {
+        return caseDetailsEntities.stream()
+                .collect(Collectors.toMap(CaseDetailsEntity::getId, Function.identity()));
+    }
+
+    private boolean hasMissingDocuments(CaseDetailsEntity caseDetails, CaseAuditEventEntity caseEvent) {
+        List<String> eventDocValues = caseEvent.getData().findValuesAsText(DOCUMENT_URL);
+        List<String> caseDocValues = caseDetails.getData().findValuesAsText(DOCUMENT_URL);
+        return !caseDocValues.containsAll(eventDocValues);
```

Comment:
`caseDocValues.containsAll(eventDocValues)`:
there could be false positives here. For example imagine `docFieldA` was wiped out, then the user after the bug period re uploads a doc to `docFieldA` but with a different name. The user has manually fixed the issue itself but we are marking the case as `hasMissingDocuments` because the old name doesn't match the new one.

Reply:
I think we should compare the event not with the case data but with the last event in the bug timeframe. That'll give us more accurate results

Reply:
Distinguishing manually corrected cases in a collection is very difficult, probably not possible. Where as in simple fields it's possible and we can treat presence of a value at that path means it's corrected or not missed at all. 
The more i think about event states more cases are coming up 😄

Reply:
> I think we should compare the event not with the case data but with the last event in the bug timeframe. That'll give us more accurate results
Actually `document_url` is unique, so does it matter?

Reply:
what I'm saying is that if you compare the event with the case, you will generate false positives also for simple fields. We can and should avoid that. We can do by comparing first event of bug timeframe with the last one rather then with the case